js/modules/modules.go

```diff
@@ -69,31 +65,108 @@ type HasModuleInstancePerVU interface {
 	NewModuleInstancePerVU() interface{}
 }
 
+// IsModuleV2 ... TODO better name
+type IsModuleV2 interface {
+	NewModuleInstance(InstanceCore) Instance
+}
+
 // checks that modules implement HasModuleInstancePerVU
 // this is done here as otherwise there will be a loop if the module imports this package
 var _ HasModuleInstancePerVU = http.New()
 
-// GetJSModules returns a map of all js modules
+// GetJSModules returns a map of all registered js modules
 func GetJSModules() map[string]interface{} {
-	result := map[string]interface{}{
-		"k6":             k6.New(),
-		"k6/crypto":      crypto.New(),
-		"k6/crypto/x509": x509.New(),
-		"k6/data":        data.New(),
-		"k6/encoding":    encoding.New(),
-		"k6/net/grpc":    grpc.New(),
-		"k6/html":        html.New(),
-		"k6/http":        http.New(),
-		"k6/metrics":     metrics.New(),
-		"k6/ws":          ws.New(),
-	}
-
 	mx.Lock()
 	defer mx.Unlock()
+	result := make(map[string]interface{}, len(modules))
 
 	for name, module := range modules {
 		result[name] = module
 	}
 
 	return result
 }
+
+// Instance is what a module needs to return
+type Instance interface {
+	InstanceCore
+	GetExports() Exports
+}
+
+func getInterfaceMethods() []string {
+	var t Instance
+	T := reflect.TypeOf(&t).Elem()
+	result := make([]string, T.NumMethod())
+
+	for i := range result {
+		result[i] = T.Method(i).Name
+	}
+
+	return result
+}
+
+// InstanceCore is something that will be provided to modules and they need to embed it in ModuleInstance
+type InstanceCore interface {
+	GetContext() context.Context
+
+	// GetInitEnv returns common.InitEnvironment instance if present
+	GetInitEnv() *common.InitEnvironment
+
+	// GetState returns lib.State if any is present
+	GetState() *lib.State
+
+	// GetRuntime returns the goja.Runtime for the current VU
+	GetRuntime() *goja.Runtime
+
+	// sealing field will help probably with pointing users that they just need to embed this in their Instance
+	// implementations
+}
+
+// Exports is representation of ESM exports of a module
+type Exports struct {
+	// Default is what will be the `default` export of a module
+	Default interface{}
+	// Named is the named exports of a module
+	Named map[string]interface{}
```

Comment:
@MStoykov I'm having an issue with exports for `k6/x/execution`, see [here](https://github.com/grafana/xk6-execution/pull/2/files#diff-1dc74514874c10c685d1ac9d2fe9f738853388a531c3b277c223e9561baad3bfR315).

With `--compatibility-mode=extended` and ES6's `import` everything works as expected. But with `--compatibility-mode=base` and ES5's `require()`, `default` is not exported... by default, and I need to manually reference it. I'm not defining `Named` since the module doesn't export any methods and everything should work via properties of the proxy.

I'm not sure if this is acceptable when working with ES5, or if the way we handle exports here should take that into account. Maybe `Named` should also be `interface{}` and we'd need more of `common.Bind`'s reflection to make it work?

Reply:
Okay some explanation on commonjs vs ESM: 

In the case of commonjs, `modules.exports` (or just `exports`) is what is returned when you do `require("modulename")` and ... that is that. The major magic is that if you set something to `modules.exports`, the alias `exports` will magically be set to the same value as it's somekind of an alias. Other than that `require` acts exactly as a function would. And this was somewhat what we just did before that ... we just returned one big object which was just returned back to the script.

ESM on the other hand is ... well not at all the same. Not only is there no object to set values to, it's all special syntax and it has this concept of default and not default(named) exports, which also have a separate way of importing.

It just so happens that babel will try to guess based on the syntax and what it gets from `require` whether it has to use `default` or `named` exports somewhere and also when it sees `export something` it will set `modules.exports` accordingly including with the `__esModule` field. So the code in this PR just emulates what babel does when it exports on its own as much as possible and allows there to be a difference between what is default and not default export.

I am not really certain what a good solution will look like especially as I have mentioned numerous times that compatibility-mode=base is now less and less useful and hopefully one day when we have import/export syntax natively we can make it even less relevant.

I did have some worries about the usage of compatibility-mode=base myself, but as I used to set `Default` and `Named` to the same thing for `k6/metrics` it works the same as before.

I guess the easiest solution is to if `Named` is nil to just put `default` as the whole return and not set `__esModule`. (and maybe the reverse if `Default` is nil :shrug: ). This will later need to change significantly again once we know what is `require` and what is `import ... from ...`.  

I did wonder briefly if I should make some Proxy/DynamicObject which if it gets `__esModule` accessed change its form so it can act as both ... but as it wasn't needed I decided against it ... but I guess it will work :shrug: .

I think the Named check is probably going to be all we need though :thinking: , can you try it @imiric ?

Reply:
Thanks for the explanation. We should mention some of that in the comment next to `__esModule` :wink:

You're right, that seems like the simplest fix for now, see da1fb43c. Works fine for `k6/x/execution` :+1: